In [1]:
# importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os


In [2]:
# installing necessary libraries as our analysis will be library specific, will be using plotly for visualization 
# and prophet for time series forecasting
#!pip install plotly


In [3]:
#!pip install prophet


In [4]:
os.getcwd()

'd:\\Data Science\\ML\\Projects\\Covid capstone project'

In [5]:
#Changing the directory to desired location
#os.chdir(r'D:\Data Science\ML\Projects\Covid capstone project')

In [6]:
os.getcwd()

'd:\\Data Science\\ML\\Projects\\Covid capstone project'

In [7]:
#importing dataset
df = pd.read_csv("D:\Data Science\ML\Projects\Covid capstone project\covid_19_clean_complete\covid_19_clean_complete.csv")

**Data Cleaning and Understanding it**

In [8]:
df

Province/State         Country/Region        Lat       Long        Date  \
0                NaN            Afghanistan  33.939110  67.709953  2020-01-22   
1                NaN                Albania  41.153300  20.168300  2020-01-22   
2                NaN                Algeria  28.033900   1.659600  2020-01-22   
3                NaN                Andorra  42.506300   1.521800  2020-01-22   
4                NaN                 Angola -11.202700  17.873900  2020-01-22   
...              ...                    ...        ...        ...         ...   
49063            NaN  Sao Tome and Principe   0.186400   6.613100  2020-07-27   
49064            NaN                  Yemen  15.552727  48.516388  2020-07-27   
49065            NaN                Comoros -11.645500  43.333300  2020-07-27   
49066            NaN             Tajikistan  38.861000  71.276100  2020-07-27   
49067            NaN                Lesotho -29.610000  28.233600  2020-07-27   

       Confirmed  Deaths  Recovered  Active             WHO Region  
0              0       0          0       0  Eastern Mediterranean  
1              0       0          0       0                 Europe  
2              0       0          0       0                 Africa  
3              0       0          0       0                 Europe  
4              0       0          0       0                 Africa  
...          ...     ...        ...     ...                    ...  
49063        865      14        734     117                 Africa  
49064       1691     483        833     375  Eastern Mediterranean  
49065        354       7        328      19                 Africa  
49066       7235      60       6028    1147                 Europe  
49067        505      12        128     365                 Africa  

[49068 rows x 10 columns]

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49068 entries, 0 to 49067
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Province/State  14664 non-null  object 
 1   Country/Region  49068 non-null  object 
 2   Lat             49068 non-null  float64
 3   Long            49068 non-null  float64
 4   Date            49068 non-null  object 
 5   Confirmed       49068 non-null  int64  
 6   Deaths          49068 non-null  int64  
 7   Recovered       49068 non-null  int64  
 8   Active          49068 non-null  int64  
 9   WHO Region      49068 non-null  object 
dtypes: float64(2), int64(4), object(4)
memory usage: 3.7+ MB


In [10]:
df.describe()

Lat          Long     Confirmed         Deaths     Recovered  \
count  49068.000000  49068.000000  4.906800e+04   49068.000000  4.906800e+04   
mean      21.433730     23.528236  1.688490e+04     884.179160  7.915713e+03   
std       24.950320     70.442740  1.273002e+05    6313.584411  5.480092e+04   
min      -51.796300   -135.000000  0.000000e+00       0.000000  0.000000e+00   
25%        7.873054    -15.310100  4.000000e+00       0.000000  0.000000e+00   
50%       23.634500     21.745300  1.680000e+02       2.000000  2.900000e+01   
75%       41.204380     80.771797  1.518250e+03      30.000000  6.660000e+02   
max       71.706900    178.065000  4.290259e+06  148011.000000  1.846641e+06   

             Active  
count  4.906800e+04  
mean   8.085012e+03  
std    7.625890e+04  
min   -1.400000e+01  
25%    0.000000e+00  
50%    2.600000e+01  
75%    6.060000e+02  
max    2.816444e+06

In [11]:
df.isnull().sum()

Province/State    34404
Country/Region        0
Lat                   0
Long                  0
Date                  0
Confirmed             0
Deaths                0
Recovered             0
Active                0
WHO Region            0
dtype: int64

In [12]:
df.duplicated().sum()

0

In [13]:
df.dtypes

Province/State     object
Country/Region     object
Lat               float64
Long              float64
Date               object
Confirmed           int64
Deaths              int64
Recovered           int64
Active              int64
WHO Region         object
dtype: object

In [14]:
#number of days taken in consideration
df.Date.nunique()

188

In [15]:
#it has gap of 1 day in between the dates
df.Date.unique()

array(['2020-01-22', '2020-01-23', '2020-01-24', '2020-01-25',
       '2020-01-26', '2020-01-27', '2020-01-28', '2020-01-29',
       '2020-01-30', '2020-01-31', '2020-02-01', '2020-02-02',
       '2020-02-03', '2020-02-04', '2020-02-05', '2020-02-06',
       '2020-02-07', '2020-02-08', '2020-02-09', '2020-02-10',
       '2020-02-11', '2020-02-12', '2020-02-13', '2020-02-14',
       '2020-02-15', '2020-02-16', '2020-02-17', '2020-02-18',
       '2020-02-19', '2020-02-20', '2020-02-21', '2020-02-22',
       '2020-02-23', '2020-02-24', '2020-02-25', '2020-02-26',
       '2020-02-27', '2020-02-28', '2020-02-29', '2020-03-01',
       '2020-03-02', '2020-03-03', '2020-03-04', '2020-03-05',
       '2020-03-06', '2020-03-07', '2020-03-08', '2020-03-09',
       '2020-03-10', '2020-03-11', '2020-03-12', '2020-03-13',
       '2020-03-14', '2020-03-15', '2020-03-16', '2020-03-17',
       '2020-03-18', '2020-03-19', '2020-03-20', '2020-03-21',
       '2020-03-22', '2020-03-23', '2020-03-24', '2020-

The dataset contains 49,068 records with global COVID-19 statistics, including confirmed cases, deaths, recoveries, and active cases. The data spans diverse regions, with latitude ranging from -51.79° to 71.70° and longitude from -135.00° to 178.06°. The distribution of cases is highly skewed, with a mean of 16,884 confirmed cases, but a median of just 168, indicating that a few locations (hotspots) had extremely high cases while most regions had relatively low numbers.

The data also shows large variations in mortality and recovery rates, with many locations reporting zero deaths while some had over 148,000 fatalities. The active cases column contains some negative values, suggesting possible reporting inconsistencies. The standard deviation is high for all major COVID-19 indicators, reflecting uneven case distribution worldwide. Moving forward, we will clean inconsistencies, visualize trends, and apply time series forecasting to predict future cases.

In [16]:
#conveting date to datetime format:
df['Date'] = pd.to_datetime(df['Date'])

In [17]:
df.head()

Province/State Country/Region       Lat       Long       Date  Confirmed  \
0            NaN    Afghanistan  33.93911  67.709953 2020-01-22          0   
1            NaN        Albania  41.15330  20.168300 2020-01-22          0   
2            NaN        Algeria  28.03390   1.659600 2020-01-22          0   
3            NaN        Andorra  42.50630   1.521800 2020-01-22          0   
4            NaN         Angola -11.20270  17.873900 2020-01-22          0   

   Deaths  Recovered  Active             WHO Region  
0       0          0       0  Eastern Mediterranean  
1       0          0       0                 Europe  
2       0          0       0                 Africa  
3       0          0       0                 Europe  
4       0          0       0                 Africa

In [18]:
# approach to compare global data with India data is adopted so creating corresponding dataframes

# Aggregate global cases
df_global = df.groupby('Date')[['Confirmed', 'Deaths', 'Recovered', 'Active']].sum().reset_index()

# Filter India-specific data
df_india = df[df['Country/Region'] == 'India'].groupby('Date')[['Confirmed', 'Deaths', 'Recovered', 'Active']].sum().reset_index()


In [19]:
df_global

Date  Confirmed  Deaths  Recovered   Active
0   2020-01-22        555      17         28      510
1   2020-01-23        654      18         30      606
2   2020-01-24        941      26         36      879
3   2020-01-25       1434      42         39     1353
4   2020-01-26       2118      56         52     2010
..         ...        ...     ...        ...      ...
183 2020-07-23   15510481  633506    8710969  6166006
184 2020-07-24   15791645  639650    8939705  6212290
185 2020-07-25   16047190  644517    9158743  6243930
186 2020-07-26   16251796  648621    9293464  6309711
187 2020-07-27   16480485  654036    9468087  6358362

[188 rows x 5 columns]

In [20]:
#no null values in global data
df_global.isnull().sum()

Date         0
Confirmed    0
Deaths       0
Recovered    0
Active       0
dtype: int64

In [21]:
#no duplicates in global data
df_global.duplicated().sum()

0

In [22]:
df_global.describe()

Confirmed         Deaths     Recovered        Active
count  1.880000e+02     188.000000  1.880000e+02  1.880000e+02
mean   4.406960e+06  230770.760638  2.066001e+06  2.110188e+06
std    4.757988e+06  217929.094183  2.627976e+06  1.969670e+06
min    5.550000e+02      17.000000  2.800000e+01  5.100000e+02
25%    1.121910e+05    3935.000000  6.044125e+04  5.864175e+04
50%    2.848733e+06  204190.000000  7.847840e+05  1.859759e+06
75%    7.422046e+06  418634.500000  3.416396e+06  3.587015e+06
max    1.648048e+07  654036.000000  9.468087e+06  6.358362e+06

The global dataset (df_global) consists of 188 days of cumulative COVID-19 data, showing a right-skewed distribution with significant variation across cases. The mean confirmed cases (~4.4M) is much lower than the max (~16.48M), indicating a few high-case days driving the average up. The median (~2.85M cases) suggests that more than half of the days had relatively lower case counts.

Deaths, recoveries, and active cases show similar trends, with deaths ranging from 17 to 654K, recoveries from 28 to 9.46M, and active cases from 510 to 6.35M. The high standard deviation across all metrics indicates fluctuating case numbers over time. Moving forward, we will analyze trends, visualize growth patterns, and build predictive models to forecast future case counts.









In [23]:
import plotly.express as px

# Create box plot for global confirmed cases
fig_global = px.box(df_global, y='Confirmed', title="Global COVID-19 Cases Distribution (Outliers Detection)",width=600, height=600)    


# Box plot for Deaths, Recovered, and Active cases
fig_deaths = px.box(df_global, y='Deaths', title="Global Deaths Distribution (Outliers Detection)",width=600, height=600)
fig_recovered = px.box(df_global, y='Recovered', title="Global Recovered Cases Distribution (Outliers Detection)",width=600, height=600)
fig_active = px.box(df_global, y='Active', title="Global Active Cases Distribution (Outliers Detection)",width=600, height=600)

fig_global.show()
fig_deaths.show()
fig_recovered.show()
fig_active.show()



In [24]:
df_india

Date  Confirmed  Deaths  Recovered  Active
0   2020-01-22          0       0          0       0
1   2020-01-23          0       0          0       0
2   2020-01-24          0       0          0       0
3   2020-01-25          0       0          0       0
4   2020-01-26          0       0          0       0
..         ...        ...     ...        ...     ...
183 2020-07-23    1288108   30601     817209  440298
184 2020-07-24    1337024   31358     849432  456234
185 2020-07-25    1385635   32060     885573  468002
186 2020-07-26    1435616   32771     917568  485277
187 2020-07-27    1480073   33408     951166  495499

[188 rows x 5 columns]

In [25]:
#No null or dublciates in India related data
print(df_india.isnull().sum())
print(df_india.duplicated().sum())

Date         0
Confirmed    0
Deaths       0
Recovered    0
Active       0
dtype: int64
0


In [26]:
df_india.describe()

Confirmed        Deaths      Recovered         Active
count  1.880000e+02    188.000000     188.000000     188.000000
mean   2.174652e+05   5913.994681  126509.148936   85042.090426
std    3.544779e+05   9000.138147  223808.885559  122839.589861
min    0.000000e+00      0.000000       0.000000       0.000000
25%    4.200000e+01      0.000000       3.000000      39.000000
50%    2.540650e+04    802.500000    5718.500000   18885.500000
75%    2.893375e+05   8201.000000  138203.250000  142205.750000
max    1.480073e+06  33408.000000  951166.000000  495499.000000

The **India-specific dataset (`df_india`)** contains **188 days of COVID-19 data**, showing a **steep rise in cases over time**. The **mean confirmed cases (~217K)** is significantly lower than the **max (~1.48M)**, indicating exponential growth in later stages. The **median (~25K cases)** suggests that for half the recorded days, cases were relatively low before surging.  

Deaths ranged from **0 to 33,408**, and recoveries from **0 to 951K**, highlighting **early pandemic phases with minimal cases**. The **high standard deviations** across all metrics reflect rapid case fluctuations. The **25th percentile (~42 cases, 0 deaths, 3 recoveries)** shows how India had almost no cases in the initial phase. Moving forward, we will analyze trends, compare India's growth with global trends, and forecast future case counts.

In [27]:
#looking for outliers in india data
# Box plot for India's confirmed cases
fig_india = px.box(df_india, y='Confirmed', title="India COVID-19 Cases Distribution (Outliers Detection)",width=600, height=600)


# Box plot for Deaths, Recovered, and Active cases
fig_india_deaths = px.box(df_india, y='Deaths', title="India Deaths Distribution (Outliers Detection)",width=600, height=600)
fig_india_recovered = px.box(df_india, y='Recovered', title="India Recovered Cases Distribution (Outliers Detection)",width=600, height=600)
fig_india_active = px.box(df_india, y='Active', title="India Active Cases Distribution (Outliers Detection)",width=600, height=600)

fig_india.show()
fig_india_deaths.show()
fig_india_recovered.show()
fig_india_active.show()


Global data doesn't suffer from extreme outlier condition but its opposite for in the case of india.

Options to proceed with:

Only work on global data and leave those few outliers as it is as they could be meaningful and droping them can create irregularities in timeline.

Winsorisation of indian dataset.

Or log transformation of dataset.


In [28]:
# Apply log transformation to global and India data
df_global['Log_Confirmed'] = np.log1p(df_global['Confirmed'])
df_india['Log_Confirmed'] = np.log1p(df_india['Confirmed'])

df_global['Log_Deaths'] = np.log1p(df_global['Deaths'])
df_india['Log_Deaths'] = np.log1p(df_india['Deaths'])

df_global['Log_Recovered'] = np.log1p(df_global['Recovered'])
df_india['Log_Recovered'] = np.log1p(df_india['Recovered'])

df_global['Log_Active'] = np.log1p(df_global['Active'])
df_india['Log_Active'] = np.log1p(df_india['Active'])





Using log1p() instead of log()?
Handling Zero Values:
If you apply log(x) directly on zero or negative values, it would result in an undefined value or error.

log(0) is undefined.
log(-x) is also undefined for real numbers.
But log1p(0) works fine because log(0 + 1) = log(1), which equals 0. This allows zero values (like no cases, no deaths, etc.) to be safely handled without causing errors.

Better Transformation for Small Values:
log1p(x) is better suited for small values because the transformation is more gradual compared to log(x) when x is close to zero. This can help with preserving small values, especially in datasets with many zeros or low counts.

In [29]:
print(df_global.describe())
print (df_india.describe())
print(df_global.shape)
print(df_india.shape)

          Confirmed         Deaths     Recovered        Active  Log_Confirmed  \
count  1.880000e+02     188.000000  1.880000e+02  1.880000e+02     188.000000   
mean   4.406960e+06  230770.760638  2.066001e+06  2.110188e+06      13.853487   
std    4.757988e+06  217929.094183  2.627976e+06  1.969670e+06       2.458064   
min    5.550000e+02      17.000000  2.800000e+01  5.100000e+02       6.320768   
25%    1.121910e+05    3935.000000  6.044125e+04  5.864175e+04      11.627852   
50%    2.848733e+06  204190.000000  7.847840e+05  1.859759e+06      14.862274   
75%    7.422046e+06  418634.500000  3.416396e+06  3.587015e+06      15.819933   
max    1.648048e+07  654036.000000  9.468087e+06  6.358362e+06      16.617688   

       Log_Deaths  Log_Recovered  Log_Active  
count  188.000000     188.000000  188.000000  
mean    10.769904      12.566978   13.319982  
std      2.742894       3.134648    2.277009  
min      2.890372       3.367296    6.236370  
25%      8.277712      11.009360   

In [30]:



# Log-transformed global data box plots (vertical orientation)
fig_log_global_confirmed = px.box(df_global, y='Log_Confirmed', title="Log Transformed Global COVID-19 Confirmed Cases",width=600, height=600)
fig_log_global_confirmed.update_layout(yaxis_title="Log (Confirmed Cases+1)")
fig_log_global_confirmed.show()

fig_log_global_deaths = px.box(df_global, y='Log_Deaths', title="Log Transformed Global COVID-19 Deaths",width=600, height=600)
fig_log_global_deaths.update_layout(yaxis_title="Log(Deaths+1)")
fig_log_global_deaths.show()

fig_log_global_recovered = px.box(df_global, y='Log_Recovered', title="Log Transformed Global COVID-19 Recovered Cases",width=600, height=600)
fig_log_global_recovered.update_layout(yaxis_title="Log (Recovered Cases+1)")
fig_log_global_recovered.show()

fig_log_global_active = px.box(df_global, y='Log_Active', title="Log Transformed Global COVID-19 Active Cases",width=600, height=600)
fig_log_global_active.update_layout(yaxis_title="Log (Active Cases+1)")
fig_log_global_active.show()


In [31]:
# Log-transformed India data box plots (vertical orientation)
fig_log_india_confirmed = px.box(df_india, y='Log_Confirmed', title="Log Transformed India COVID-19 Confirmed Cases",width=600, height=600)
fig_log_india_confirmed.update_layout(yaxis_title="Log (Confirmed Cases+1)")
fig_log_india_confirmed.show()

fig_log_india_deaths = px.box(df_india, y='Log_Deaths', title="Log Transformed India COVID-19 Deaths",width=600, height=600)
fig_log_india_deaths.update_layout(yaxis_title="Log(Deaths+1)")
fig_log_india_deaths.show()

fig_log_india_recovered = px.box(df_india, y='Log_Recovered', title="Log Transformed India COVID-19 Recovered Cases",width=600, height=600)
fig_log_india_recovered.update_layout(yaxis_title="Log (Recovered Cases+1)")
fig_log_india_recovered.show()

fig_log_india_active = px.box(df_india, y='Log_Active', title="Log Transformed India COVID-19 Active Cases",width=600, height=600)
fig_log_india_active.update_layout(yaxis_title="Log (Active Cases+1)")
fig_log_india_active.show()


In [32]:
#Forecasting using prophet


# Prepareing the data for Prophet (log-transformed confirmed cases)
df_global_prophet = df_global[['Date', 'Log_Confirmed']].rename(columns={'Date': 'ds', 'Log_Confirmed': 'y'})
df_india_prophet = df_india[['Date', 'Log_Confirmed']].rename(columns={'Date': 'ds', 'Log_Confirmed': 'y'})



In [33]:
from prophet import Prophet

# Initializing Prophet models
model_global = Prophet()
model_india = Prophet()

# Fit the models to the data
model_global.fit(df_global_prophet)
model_india.fit(df_india_prophet)


00:38:26 - cmdstanpy - INFO - Chain [1] start processing
00:38:27 - cmdstanpy - INFO - Chain [1] done processing
00:38:27 - cmdstanpy - INFO - Chain [1] start processing
00:38:27 - cmdstanpy - INFO - Chain [1] done processing


In [34]:
# Prophet model for Global and India without passing dataframe directly
future_global = model_global.make_future_dataframe(periods=7)
future_india = model_india.make_future_dataframe(periods=7)



In [35]:
# Generating predictions for the global and India datasets
forecast_global = model_global.predict(future_global)
forecast_india = model_india.predict(future_india)


In [36]:
import plotly.graph_objects as go

# Creating a Plotly figure for Global data
fig_global_forecast = go.Figure()

# Adding predicted data (forecasted log-transformed cases)
fig_global_forecast.add_trace(go.Scatter(x=forecast_global['ds'], y=forecast_global['yhat'],
                                        mode='lines', name='Predicted Data', line=dict(color='orange', dash='dash')))

# Adding confidence intervals
fig_global_forecast.add_trace(go.Scatter(x=forecast_global['ds'], y=forecast_global['yhat_lower'],
                                        mode='lines', name='Lower Confidence Interval', line=dict(color='grey', dash='dot')))
fig_global_forecast.add_trace(go.Scatter(x=forecast_global['ds'], y=forecast_global['yhat_upper'],
                                        mode='lines', name='Upper Confidence Interval', line=dict(color='grey', dash='dot')))

# Adding actual data (log-transformed confirmed cases)
fig_global_forecast.add_trace(go.Scatter(x=df_global_prophet['ds'], y=df_global_prophet['y'],
                                        mode='lines', name='Actual Data', line=dict(color='blue')))

# Updating layout for better readability
fig_global_forecast.update_layout(
    title="Global COVID-19 Log-Confirmed Cases Forecast",
    title_font_size=18,
    xaxis_title="Date",
    yaxis_title="Log(Confirmed Cases + 1)",  # Correct label indicating log-transformation
    xaxis=dict(tickangle=45),  # Rotate X-axis labels for better visibility
    yaxis=dict(showgrid=True),
    font=dict(size=14),
    width=1200, height=600
)

# Show the plot
fig_global_forecast.show()


In [37]:


# Creating a Plotly figure for India data
fig_india_forecast = go.Figure()

# Adding predicted data (forecasted log-transformed cases)
fig_india_forecast.add_trace(go.Scatter(x=forecast_india['ds'], y=forecast_india['yhat'],
                                       mode='lines', name='Predicted Data', line=dict(color='orange', dash='dash')))

# Adding confidence intervals
fig_india_forecast.add_trace(go.Scatter(x=forecast_india['ds'], y=forecast_india['yhat_lower'],
                                       mode='lines', name='Lower Confidence Interval', line=dict(color='grey', dash='dot')))
fig_india_forecast.add_trace(go.Scatter(x=forecast_india['ds'], y=forecast_india['yhat_upper'],
                                       mode='lines', name='Upper Confidence Interval', line=dict(color='grey', dash='dot')))

# Adding actual data (log-transformed confirmed cases)
fig_india_forecast.add_trace(go.Scatter(x=df_india_prophet['ds'], y=df_india_prophet['y'],
                                       mode='lines', name='Actual Data', line=dict(color='blue')))

# Updating layout for better readability
fig_india_forecast.update_layout(
    title="India COVID-19 Log-Confirmed Cases Forecast",
    title_font_size=18,
    xaxis_title="Date",
    yaxis_title="Log(Confirmed Cases + 1)",  # Correct label indicating log-transformation
    xaxis=dict(tickangle=45),  # Rotate X-axis labels for better visibility
    yaxis=dict(showgrid=True),
    font=dict(size=14),
    width=1200, height=600
)

# Show the plot
fig_india_forecast.show()


In [38]:
import plotly.graph_objects as go

# Creating a Plotly figure for Global Recovered cases (using actual data)
fig_global_recovered = go.Figure()

# Adding actual recovered data from the original dataset
fig_global_recovered.add_trace(go.Scatter(x=df_global['Date'], y=df_global['Recovered'],
                                         mode='lines', name='Actual Recovered', line=dict(color='darkgreen')))

# Updating layout for better readability
fig_global_recovered.update_layout(
    title="Global COVID-19 Recovered Cases (Actual Data)",
    title_font_size=18,
    xaxis_title="Date",
    yaxis_title="Recovered Cases",
    xaxis=dict(tickangle=45),  # Rotate X-axis labels for better visibility
    yaxis=dict(showgrid=True),
    font=dict(size=14),
    width=1500, height=600
)

# Show the plot
fig_global_recovered.show()




In [39]:
import plotly.graph_objects as go

# Creating a Plotly figure for India Recovered cases (using actual data)
fig_india_recovered = go.Figure()

# Adding actual recovered data from the original dataset
fig_india_recovered.add_trace(go.Scatter(x=df_india['Date'], y=df_india['Recovered'],
                                        mode='lines', name='Actual Recovered', line=dict(color='darkgreen')))

# Updating layout for better readability
fig_india_recovered.update_layout(
    title="India COVID-19 Recovered Cases (Actual Data)",
    title_font_size=18,
    xaxis_title="Date",
    yaxis_title="Recovered Cases",
    xaxis=dict(tickangle=45),  # Rotate X-axis labels for better visibility
    yaxis=dict(showgrid=True),
    font=dict(size=14),
    width=1500, height=600
)

# Show the plot
fig_india_recovered.show()


**Testing Model Performance**

***For India Specific dataset***

In [40]:
from sklearn.metrics import mean_absolute_error, mean_squared_error


# Splitting into training (first 160 days) and testing (last 28 days)
train = df_india_prophet.iloc[:160]
test = df_india_prophet.iloc[160:]

In [41]:
df_india_prophet.shape, train.shape, test.shape

((188, 2), (160, 2), (28, 2))

In [42]:
# Initialize Prophet model
model_india = Prophet()

# Fit the model on training data
model_india.fit(train)

# Create future dates matching test period (28 days ahead)
future_india = model_india.make_future_dataframe(periods=28)


00:38:33 - cmdstanpy - INFO - Chain [1] start processing
00:38:33 - cmdstanpy - INFO - Chain [1] done processing


In [43]:
# Generate forecast
forecast_india = model_india.predict(future_india)

# Extract predictions only for test period
forecast_test = forecast_india[['ds', 'yhat']].merge(test[['ds', 'y']], on='ds')

# Display actual vs predicted values
print(forecast_test.head())


          ds       yhat          y
0 2020-06-30  13.307211  13.280191
1 2020-07-01  13.406933  13.312392
2 2020-07-02  13.458923  13.346379
3 2020-07-03  13.466686  13.382134
4 2020-07-04  13.489990  13.419747


Mean Absolute Percentage Error (MAPE)

In [44]:
def mean_absolute_percentage_error(y_true, y_pred):
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

mape = mean_absolute_percentage_error(forecast_test['y'], forecast_test['yhat'])
print(f'MAPE: {mape:.2f}%')


MAPE: 1.01%


Root Mean Squared Error (RMSE)

In [45]:
rmse = mean_squared_error(forecast_test['y'], forecast_test['yhat'], squared=False)
print(f'RMSE: {rmse:.2f}')


RMSE: 0.15


Mean Absolute Error (MAE)

In [46]:
mae = mean_absolute_error(forecast_test['y'], forecast_test['yhat'])
print(f'MAE: {mae:.2f}')


MAE: 0.14


In [47]:
import plotly.graph_objects as go

# Create figure
fig = go.Figure()

# Add actual values (test data)
fig.add_trace(go.Scatter(
    x=test['ds'], 
    y=test['y'], 
    mode='lines+markers', 
    name='Actual', 
    line=dict(color='blue')
))

# Add predicted values (forecasted data)
fig.add_trace(go.Scatter(
    x=forecast_test['ds'], 
    y=forecast_test['yhat'], 
    mode='lines+markers', 
    name='Predicted', 
    line=dict(color='red', dash='dot')
))

# Customize layout
fig.update_layout(
    title="Actual vs Predicted Values for India",
    xaxis_title="Date",
    yaxis_title="Value",
    legend=dict(x=0, y=1),
    template="plotly_dark"
)

# Show figure
fig.show()


***For Globally representative data***

In [48]:
# Split data: First 160 days for training, last 28 days for testing
train_global = df_global_prophet.iloc[:160]
test_global = df_global_prophet.iloc[160:]

# Initialize Prophet model
model_global = Prophet()

# Fit the model on training data
model_global.fit(train_global)

# Create future dates matching test period (28 days ahead)
future_global = model_global.make_future_dataframe(periods=28)


00:38:34 - cmdstanpy - INFO - Chain [1] start processing
00:38:34 - cmdstanpy - INFO - Chain [1] done processing


In [49]:
# Generate forecast
forecast_global = model_global.predict(future_global)

# Extract predictions only for test period
forecast_test_global = forecast_global[['ds', 'yhat']].merge(test_global[['ds', 'y']], on='ds')

# Display actual vs predicted values
print(forecast_test_global.head())


          ds       yhat          y
0 2020-06-30  16.176191  16.162084
1 2020-07-01  16.186650  16.182702
2 2020-07-02  16.201750  16.201986
3 2020-07-03  16.220383  16.220525
4 2020-07-04  16.233830  16.237846


Mean Absolute Percentage Error (MAPE)

In [50]:
mape_global = mean_absolute_percentage_error(forecast_test_global['y'], forecast_test_global['yhat'])
print(f'MAPE: {mape_global:.2f}%')

MAPE: 0.06%


Root Mean Squared Error (RMSE)

In [51]:
rmse_global = mean_squared_error(forecast_test_global['y'], forecast_test_global['yhat'], squared=False)
print(f'RMSE: {rmse_global:.2f}')


RMSE: 0.01


Mean Absolute Error (MAE)

In [52]:
mae_global = mean_absolute_error(forecast_test_global['y'], forecast_test_global['yhat'])
print(f'MAE: {mae_global:.2f}')


MAE: 0.01


In [53]:
# Create figure
fig = go.Figure()

# Add actual values (test data)
fig.add_trace(go.Scatter(
    x=test_global['ds'], 
    y=test_global['y'], 
    mode='lines+markers', 
    name='Actual', 
    line=dict(color='blue')
))

# Add predicted values (forecasted data)
fig.add_trace(go.Scatter(
    x=forecast_test_global['ds'], 
    y=forecast_test_global['yhat'], 
    mode='lines+markers', 
    name='Predicted', 
    line=dict(color='red', dash='dot')
))

# Customize layout
fig.update_layout(
    title="Actual vs Predicted Values for Global Data",
    xaxis_title="Date",
    yaxis_title="Value",
    legend=dict(x=0, y=1),
    template="plotly_dark"
)

# Show figure
fig.show()
